src/middlewared/middlewared/plugins/iscsi_/alua.py

```diff
@@ -0,0 +1,390 @@
+import asyncio
+import itertools
+
+from middlewared.service import Service, job
+
+CHUNK_SIZE = 20
+RETRY_SECONDS = 5
+HA_TARGET_SETTLE_SECONDS = 10
+GET_UNIT_STATE_SECONDS = 2
+
+
+class iSCSITargetAluaService(Service):
+    """
+    Support iSCSI ALUA configuration.
+
+    The ALUA mechanism is based up DLM support baked into SCST, along with other
+    potions of middleware (dlm, iscsi.targets, etc ) to handle the coordination
+    between the two nodes in a HA pair.  This is performed in response to
+    cluster_mode being set on target extents.
+
+    However, when a LARGE number of extents(/targets) are present it becomes
+    impractical to leave/enter lockspaces on scst startup.
+
+    To avoid this, SCST on the ACTIVE will start without cluster_mode being
+    set on extents.  Likewise on the STANDBY node, so the targets there will be
+    present but disabled.  However, the STANDBY will then initiate a job to
+    (gradually) enable cluster_mode on the ACTIVE and react.
+    """
+    class Config:
+        private = True
+        namespace = 'iscsi.alua'
+
+    # See HA_PROPAGATE in event.py.  Not only required when running command
+    # on MASTER, and don't want it to propagate.
+    HA_PROPAGATE = {'ha_propagate': False}
+
+    def __init__(self, middleware):
+        super().__init__(middleware)
+        self.enabled = set()
+        self.standby_starting = False
+
+        self.active_elected_job = None
+        self.activate_extents_job = None
+
+        # scst.conf.mako will check the below value to determine whether it
+        # should write cluster_mode = 1, even if then value in /sys is zero
+        # (i.e. extent not listed in iscsi.target.clustered_extents)
+        self._all_cluster_mode = False
+
+        # Likewise
+        self._standby_write_empty_config = True
+
+    async def before_start(self):
+        if await self.middleware.call('iscsi.global.alua_enabled'):
+            if await self.middleware.call('failover.status') == 'BACKUP':
+                self._standby_write_empty_config = True
+                await self.middleware.call('etc.generate', 'scst')
+
+    async def after_start(self):
+        if await self.middleware.call('iscsi.global.alua_enabled'):
+            if await self.middleware.call('failover.status') == 'BACKUP':
+                await self.middleware.call('iscsi.alua.standby_after_start')
+
+    async def before_stop(self):
+        self._all_cluster_mode = False
+        self.standby_starting = False
+
+    async def standby_enable_devices(self, devices):
+        await self.middleware.call('iscsi.target.login_ha_targets')
+        extents = await self.middleware.call('iscsi.extent.logged_in_extents')
+        asked = set(devices)
```

Comment:
The `asked` variable isn't used anywhere?

Reply:
Used on next line.